370-auth-ui/oidc.md

```diff
@@ -0,0 +1,455 @@
+
+`response_type=code&scope=openid&client_id=CLIENT_ID&redirect_uri=REDIRECT_URI&code_challenge_method=S256&code_challenge=CODE_CHALLENGE&ui_locales=en`
+
+### Token Request
+
+The following sessions list out the request parameters and their difference from the spec.
+
+#### grant_type
+
+Either `authorization_code` or `refresh_token`
+
+#### code
+
+No difference from the spec.
+
+#### redirect_uri
+
+No difference from the spec.
+
+#### client_id
+
+It is always required. The value must be a valid Skygear API Key.
+
+#### code_verifier
+
+No difference from the spec.
+
+#### Example of valid Token Request
+
+`grant_type=authorization_code&code=AUTHORIZATION_CODE&redirect_uri=REDIRECT_URI&client_id=API_KEY&code_verifier=CODE_VERIFIER`
+
+### Token Response
+
+#### id_token
+
+No difference from the spec.
+
+#### token_type
+
+It is always the value `bearer`.
+
+#### access_token
+
+No difference from the spec.
+
+#### expires_in
+
+No difference from the spec.
+
+#### refresh_token
+
+No difference from the spec. Absent in [SPA](#case-2-single-page-web-app-spa) flow.
+
+#### scope
+
+It is always absent.
+
+#### Example of Token Response
+
+```json
+{
+  "token_type": "bearer",
+  "access_token": "access.token",
+  "refresh_token": "refresh.token"
+}
+```
+
+## Authorization Server Metadata Document
+
+Skygear supports OpenID Connect and OAuth metadata document which contains most of the information required for an app to do sign-in. The document includes information about endpoints and location of public keys. It is in JSON format, for values definition see [OpenID Connect Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) and [Authorization Server Metadata](https://tools.ietf.org/html/rfc8414#section-2).
+
+The OpenID Connect discovery document can be retrieved from:
+
+```
+https://{auth_gear_endpoint}/.well-known/openid-configuration
+```
+
+The OAuth Authorization Server Metadata can be retrieved from:
+
+```
+https://{auth_gear_endpoint}/.well-known/oauth-authorization-server
+```
+
+The following sessions list out the metadata values that Skygear Auth supports.
+
+#### issuer
+
+No difference from the spec.
+
+#### authorization_endpoint
+
+No difference from the spec.
+
+#### token_endpoint
+
+No difference from the spec.
+
+#### userinfo_endpoint
+
+No difference from the spec.
+
+#### jwks_uri
+
+No difference from the spec.
+
+### scopes_supported
+
+The values will be `["openid", "offline_access"]`.
+
+#### response_types_supported`
+
+The values will be `["code", "none"]`.
+
+#### grant_types_supported
+
+The values will be `["authorization_code", "refresh_token"]`.
+
+#### subject_types_supported
+
+The values will be `["public"]`.
+
+#### id_token_signing_alg_values_supported
+
+`RS256` is supported.
+
+#### claims_supported
+
+The values will be `["sub", "iss", "aud", "exp", "iat"]`.
+
+### code_challenge_methods_supported
+
+The values will be `["plain", "S256"]`
```

Comment:
We only support `S256`.

Reply:
But to be fully compliant with the spec, we must support `plain`.

Reply:
I think PKCE spec does not require support of `plain` transformation. Quoting from spec:
- 4.2: 'Clients are permitted to use "plain" only if they cannot support "S256" for some technical reason and know via out-of-band configuration that the server supports "plain".'
- 4.4.1: 'If the server supporting PKCE does not support the requested
   transformation, the authorization endpoint MUST return the
   authorization error response with "error" value set to
   "invalid_request".'

This shows the spec allow server to return error response in case "plain" transformation is not supported.

Reply:
Maybe update to support `S256` first? And we will only use `S256` actually :)